Generalize dependencies_to_sdsmj() and gitlog_to_hdsmj() in transform_*() and graph_to_dsmj()

leilani-reich opened the pull request:
I will update the graph.R and network.R with the graph_to_dsmj(), transform_dependencies_to_sdsmj(), and transform_gitlog_to_hdsmj() functions for milestone 3.4 of DV8 integration.

Right now, I have done the following:
- replaced use of accessing columns by indices (using the column name instead)
- replaced use of data.frame (replaced with as.data.table)
- new parameters dsm_path, hdsmj_path, and sdsmj_path for my 3 functions
- updated the documentation of my 3 functions
- added new parameter "is_directed" for my 3 functions, with no default value. If is_directed is false, the cells will be doubled in graph_to_dsmj() with the src/dest names reversed
- tested the use of
 ```r
cochange_table <- transform_gitlog_to_temporal_network(project_git, mode=c("author", "committer"))
```

instead of 

 ```r
  git_bipartite <- transform_gitlog_to_bipartite_network(project_git, mode ="commit-file")
  cochange_table <- bipartite_graph_projection(git_bipartite, mode = FALSE,
                                               weight_scheme_function = weight_scheme_count_deleted_nodes)
```

in my transform_gitlog_to_hdsmj().

I tried this on my calculator project and got the following json.
[april27-calc-gitlog.json.zip](https://github.com/sailuh/kaiaulu/files/11352956/april27-calc-gitlog.json.zip)

I am noticing some cells where src/dest are the same, which matches the nodes with arrows pointing to themselves in the temporal network [here](http://itm0.shidler.hawaii.edu/kaiaulu/articles/gitlog_showcase.html#bipartite-networks-and-projections). I also notice the graph looks directed in terms of the cell output.

Is this what you wanted me to look at for the generalization? I also ask on the issue [here](https://github.com/sailuh/kaiaulu/issues/184#issuecomment-1527253324).

And do you want the transform_gitlog_to_hdsmj() to have the option to choose the temporal network? And if so, what should I rename the "Cochange" value to in my hdsm.json?

Thanks

carlosparadis commented on R/graph.R:
```diff
@@ -7,19 +7,24 @@
+#'  with src/dest switched to represent an undirected graph as a directed graph.
+#' @param dsm_path path to save either the *-sdsm.json from \code{\link{transform_dependencies_to_sdsmj}} or the
+#' *-hdsmj.json from \code{\link{transform_gitlog_to_hdsmj}}
 #' @param is_sorted whether to sort the variables (filenames) in the dsm.json files
```
add to the end of this sentence `(optional)`

carlosparadis commented on R/graph.R:
```diff
@@ -7,19 +7,24 @@
+#' @param dsm_path path to save either the *-sdsm.json from \code{\link{transform_dependencies_to_sdsmj}} or the
+#' *-hdsmj.json from \code{\link{transform_gitlog_to_hdsmj}}
 #' @param is_sorted whether to sort the variables (filenames) in the dsm.json files
+#' @return either the path to a sdsm.json (from dependencies table) or a hdsmj.json (from gitlog table).
```
@return the path to the dsmj.json file saved

carlosparadis commented on R/graph.R:
```diff
@@ -7,19 +7,24 @@
   # Get the nodes and edgelist tables
-  nodes_table <- graph[[1]]
-  edgelist_table <- graph[[2]]
+  nodes_table <- graph[["nodes"]]
```
yes, good!

carlosparadis commented on R/graph.R:
```diff
@@ -53,28 +58,56 @@ graph_to_dsmj <- function(graph, is_sorted=FALSE){
 
   # Create the final json
-  hdsm_json <- list(schemaVersion="1.0", name=paste0("april16-graph", "-hdsm"), variables=variables, cells=cells_df)
+  hdsm_json <- list(schemaVersion="1.0", name=paste0("hdsm"), variables=variables, cells=cells_df)
```
this should just be dsm_json (no prefix h). Moreover, you should not try to suffix or format the file name. Assume the user pass as parameter the full path location including file name and extension. 

I am not sure if dV8 cares for this parameter, but if necessary, add one more parameter to graph_to_dsmj and pass the "hdsm" via the transform function for now. Any assumptions should be defer to the transforms.

carlosparadis commented on R/graph.R:
```diff
@@ -53,28 +58,56 @@ graph_to_dsmj <- function(graph, is_sorted=FALSE){
-  hdsm_json <- list(schemaVersion="1.0", name=paste0("april16-graph", "-hdsm"), variables=variables, cells=cells_df)
+  hdsm_json <- list(schemaVersion="1.0", name=paste0("hdsm"), variables=variables, cells=cells_df)
 
   json_df <- jsonlite::fromJSON(jsonlite::toJSON(hdsm_json, auto_unbox = TRUE))
```
make sure to rename these variables name to dsm_json

carlosparadis commented on R/network.R:
```diff
@@ -6,46 +6,69 @@
   # Rename "src_filepath" to "from"
-  colnames(depends_table$edgelist)[1] = "from"
+  colnames(depends_table$edgelist)[colnames(depends_table$edgelist) == "src_filepath"] = "from"
   # Rename "dest_filepath" to "to"
-  colnames(depends_table$edgelist)[2] = "to"
+  colnames(depends_table$edgelist)[colnames(depends_table$edgelist) == "dest_filepath"] = "to"
```
Use:

setnames(x=depends_table[["edgelist"]],
                  old = c("src_filepath","dest_filepath") 
                  to = c("from","to")                             
                  )
This should not require an assignment. It is one of the peculiarities of data.table, and why it makes it so fast.

carlosparadis commented on R/network.R:
```diff
@@ -6,46 +6,69 @@
+transform_dependencies_to_sdsmj <- function(project_depends, sdsmj_path, is_directed, is_sorted=FALSE){
   # Rename "filepath" to "to"
-  colnames(depends_table$nodes)[1] = "name"
+  colnames(depends_table$nodes)[colnames(depends_table$nodes) == "filepath"] = "name"
```
Use setnames here too.

carlosparadis reviewed (changes requested):
I made some inline corrections.

leilani-reich commented:
Hi @carlosparadis, I made some updates to my functions and added the separate transform_temporal_gitlog_to_adsmj. I also fixed how I was getting the parameters in the graph_to_dsmj. I ran the functions and they all gave what looked like expected json outputs, but I will test more thoroughly. Also, I think I am noticing issues with the parse_dependencies() function again, the filepath it takes in. I will let you know about that.

leilani-reich commented on R/parser.R:
```diff
@@ -579,6 +579,8 @@ parse_dependencies <- function(depends_jar_path,git_repo_path,language,output_di
   project_name <- stri_split_regex(folder_path,pattern="/")[[1]]
   project_name <- project_name[length(project_name)-1]
+
+  print(folder_path)
```
note to self: remember to remove

leilani-reich commented on R/graph.R:
```diff
@@ -31,92 +33,62 @@ graph_to_dsmj <- function(graph, dsmj_path, dsmj_name, is_directed, is_sorted=FA
-    src <- edgelist_table[["from"]][[i]]
-    dest <- edgelist_table[["to"]][[i]]
+  # Group data into cells by the "from" and "to" files and get their labels and weights.
+  grouped_cells <- edgelist_table[, .(labels=list(label), weights=list(weight)),
```
I use a group by so that I can get all of the labels and weights for each src/dest pair, and so i can use grouped_cells in my getCell and getCellReverse function

leilani-reich commented on R/network.R:
```diff
@@ -77,36 +69,32 @@ transform_gitlog_to_hdsmj <- function(project_git, hdsmj_path, is_sorted=FALSE){
-#'   were committed together.
-#' Note that the cochange between a file and its renamed variant will not be considered
-#' using this function, so those cells won't appear in the final *-hdsm.json.
+#'  variable (src & dest) pairs. The Collaborate value is the number of times the src author and dest author changed the same file.
```
I'm not sure if my definition of the Collaborate value is the best.

carlosparadis replied:
I think this is fine!

leilani-reich commented on R/network.R:
```diff
@@ -13,57 +13,49 @@
-  project_dependencies[["nodes"]] <- setnames(x=project_dependencies[["nodes"]],
-                                                 old = c("filepath"),
-                                                 new = c("name"))
+  setnames(x=project_dependencies[["nodes"]], old = c("filepath"), new = c("name"))
```
For some reason, once I run transform_dependencies_to_sdsmj() on my parsed dependencies_table, if I try to load my parsed_dependencies_table again then parsed_dependencies_table[["nodes"]] column called "filepath" is renamed to "name", which I did in transform_dependencies_to_sdsmj(). However, in parsed_dependencies_table[["edgelist"]], the column names are not renamed. I don't understand why this is.

carlosparadis replied:
I'm confused about what the issue is here. In the code above you are only modifying the [["nodes"]] table, so the [["edgelist"]] columns remain unaffected. Regardless, the edgelist column names should be `from` and `to` for the filepaths, not name. Did I misunderstand the question?

leilani-reich replied:
Sorry, the line below this shows where I renamed the edgelist columns.

```r
  setnames(x=project_dependencies[["edgelist"]], old = c("src_filepath","dest_filepath", "value"),
           new = c("from","to", "weight"))
```

And in regards to the nodes table, I renamed it to "name" from "filepath" since that was mentioned in my dv8 functions issue a while back https://github.com/sailuh/kaiaulu/issues/168#issuecomment-1509333620.

So to clarify the problem, I have this table
```r
parsed_depends_table <- kaiaulu::parse_dependencies(depends_path, git_repo_path, lang, output_dir)
parsed_depends_table
```

and then I call my function
```r
kaiaulu::transform_dependencies_to_sdsmj(parsed_depends_table,sdsmj_path="calc-april29-depends.json",is_sorted=TRUE)
```

and then I try to print the parsed_depends_table again
The nodes table "filepath" column is now "name"
<img width="655" alt="Screenshot 2023-04-29 at 7 55 04 PM" src="https://user-images.githubusercontent.com/89750594/235338130-1ada854d-2d12-44fe-94cc-c615f46b5aba.png">

while the edgelist table columns did not change
<img width="512" alt="Screenshot 2023-04-29 at 7 55 10 PM" src="https://user-images.githubusercontent.com/89750594/235338149-0137b4c2-b5d8-45a0-9c7f-be832a32ede9.png">

I just think this behavior is odd. Also, should the function even be modifying the original "parsed_depends_table" columns? If the user wanted to use that parsed_depends_table for something else, I think this would cause issues.

carlosparadis replied:
Your concern is valid.

If the issue is, it is modifying the original table, try looking up Google how data.table modifies objects by reference instead of copying. This is what makes data.table so efficient, aside from being implemented in C underneath. 

On your transform function, you want to copy the object and apply setnames to the copied object instead of setnames to the original. Accordingly, you also want to pass the copied object to graph_to_dsmj() too.

Have a look on this: 

https://rdrr.io/cran/data.table/man/copy.html#:~:text=data.table%20provides%20functions%20that%20operate%20on%20objects%20by,used%20to%20subassign%20to%20a%20column%20by%20reference.

Let me know if I misunderstood or it is not addressing your issue.

leilani-reich replied:
Thank you, this works!

leilani-reich commented:
## Checking output of graph_to_dsmj

I checked the output json for my transform_gitlog_to_hdsmj, transform_dependencies_to_sdsmj, and transform_temporal_gitlog_to_adsmj. From what I checked, the results look reasonable. 

For transform_gitlog_to_hdsmj and transform_dependencies_to_sdsmj, I was able to use the DV8 output for comparison and saw that it aligned.

As for transform_temporal_gitlog_to_adsmj, I'm not sure if I can verify that the "Collaborate" values are correct, but I visualized the graph following similar steps to the temporal network graph shown [here](http://itm0.shidler.hawaii.edu/kaiaulu/articles/gitlog_showcase.html#temporal-networks), and got the following:

<img width="819" alt="Screenshot 2023-04-29 at 11 07 49 PM" src="https://user-images.githubusercontent.com/89750594/235345080-b9175d51-5525-459e-9b0e-176746ee8f69.png">

This was done on the https://github.com/HouariZegai/Calculator git repo. Looking at the graph, I saw that the cells in my adsmj matched in terms of src authors having arrows pointing to dest authors.

## json output files to check

Here are my output jsons from each of my transform functions. I have also included a dv8 output hdsm to compare to my transform_gitlog_to_hdsmj output, and a dv8 output sdsm to compare to my transform_dependencies_to_sdsmj output.

[april29tests.zip](https://github.com/sailuh/kaiaulu/files/11360951/april29tests.zip)

carlosparadis reviewed (approved):
Looks good to me. I don't see any issues, and I see the revisions were implemented too. Checks are also passing. I will try using them on the DV8 Notebook.
